clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java

```diff
@@ -387,58 +403,125 @@ private boolean allSubscriptionsEqual(Set<String> allTopics,
-                    sortedConsumers.put(memberData.generation.orElse(DEFAULT_GENERATION), consumer);
-                    sortedPartitionConsumersByGeneration.put(partition, sortedConsumers);
-                }
-            }
-        }
+            List<TopicPartition> ownedPartitions = new ArrayList<>();
+            currentAssignment.put(consumer, ownedPartitions);
 
-        // prevAssignment holds the prior ConsumerGenerationPair (before current) of each partition
-        // current and previous consumers are the last two consumers of each partition in the above sorted map
-        for (Map.Entry<TopicPartition, TreeMap<Integer, String>> partitionConsumersEntry: sortedPartitionConsumersByGeneration.entrySet()) {
-            TopicPartition partition = partitionConsumersEntry.getKey();
-            TreeMap<Integer, String> consumers = partitionConsumersEntry.getValue();
-            Iterator<Integer> it = consumers.descendingKeySet().iterator();
+            // Only consider this consumer's owned partitions as valid if it is a member of the current highest
+            // generation, or it's generation is not present but we have not seen any known generation so far
+            if (memberData.generation.isPresent() && memberData.generation.get() >= maxGeneration
+                || !memberData.generation.isPresent() && maxGeneration == DEFAULT_GENERATION) {
 
-            // let's process the current (most recent) consumer first
-            String consumer = consumers.get(it.next());
-            currentAssignment.computeIfAbsent(consumer, k -> new ArrayList<>());
-            currentAssignment.get(consumer).add(partition);
+                // If the current member's generation is higher, all the previously owned partitions are invalid
+                if (memberData.generation.isPresent() && memberData.generation.get() > maxGeneration) {
+                    for (String member: membersOfCurrentHighestGeneration) {
+                        List<TopicPartition> oldGenerationPartitions = currentAssignment.get(member);
+                        updatePrevAssignment(prevAssignment, oldGenerationPartitions, member, maxGeneration);
+                        oldGenerationPartitions.clear();
+                    }
+                    membersOfCurrentHighestGeneration.clear();
+                    maxGeneration = memberData.generation.get();
+                }
 
-            // now update previous assignment if any
-            if (it.hasNext()) {
-                int generation = it.next();
-                prevAssignment.put(partition, new ConsumerGenerationPair(consumers.get(generation), generation));
+                membersOfCurrentHighestGeneration.add(consumer);
+                ownedPartitions.addAll(memberData.partitions);
+            } else if (!memberData.generation.isPresent()) {
+                // current maxGeneration is larger than DEFAULT_GENERATION,
+                // put all partitions as DEFAULT_GENERATION into provAssignment
```

Comment:
typo: prev.

Reply:
I'm not 100% sure if the refactored code has the exactly same logic as the old code now since its branching conditions have largely changed. E.g. do we still detect if a partition is assigned to different consumers in a generation or not?

Reply:
My refactor is just trying to reach the same `currentAssignment` and `prevAssignment` as before. So, if you meant this:
```java
if (memberData.generation.isPresent() && consumers.containsKey(memberData.generation.get())) {
    // same partition is assigned to two consumers during the same rebalance.
    // log a warning and skip this record
    log.warn("Partition '{}' is assigned to multiple consumers following sticky assignment generation {}.",
        partition, memberData.generation);
```
I think this check is unnecessary since we didn't do anything to it, and we cannot do anything to it, either. But that's my thought. Thanks.

Reply:
Sounds good.